encodings/zstd/src/array.rs

```diff
@@ -0,0 +1,321 @@
+            let compressed = compressor
+                .compress(uncompressed)
+                .map_err(|err| VortexError::from(err).with_context("while compressing"))?;
+            let frame_n_rows = (frame_row_indices.get(i + 1).cloned().unwrap_or(n_rows)
+                - frame_row_indices[i]) as u64;
+            frame_metas.push(ZstdFrameMetadata {
+                n_rows: frame_n_rows,
+                compressed_size: compressed.len() as u64,
+                uncompressed_size: uncompressed.len() as u64,
+            });
+            frames.push(ByteBuffer::from(compressed));
+        }
+
+        let metadata = ZstdMetadata {
+            dictionary_size: dictionary
+                .as_ref()
+                .map_or(0, |dict| dict.len())
+                .try_into()?,
+            frames: frame_metas,
+        };
+
+        Ok(ZstdArray::new(
+            dictionary, frames, dtype, metadata, n_rows, validity,
+        ))
+    }
+
+    pub fn from_array(array: ArrayRef, level: i32, rows_per_frame: usize) -> VortexResult<Self> {
+        if let Some(parray) = array.as_opt::<PrimitiveVTable>() {
+            Self::from_primitive(parray, level, rows_per_frame)
+        } else {
+            Err(vortex_err!("Zstd can only encode primitive arrays"))
+        }
+    }
+
+    pub fn decompress(&self) -> VortexResult<ArrayRef> {
+        // To start, we figure out which frames we need to decompress, and with
+        // what row offset into the first such frame.
+        let slice_n_rows = self.slice_stop - self.slice_start;
+        let byte_width = self.dtype.as_ptype().byte_width();
+        let mut frame_start_row = 0;
+        let mut frame_idx_lb = 0;
+        let mut frame_idx_ub = 0;
+        let mut row_offset = 0;
+        for (i, frame_meta) in self.metadata.frames.iter().enumerate() {
+            let buf_stop = frame_start_row + usize::try_from(frame_meta.n_rows)?;
+            if frame_start_row < self.slice_start {
+                frame_idx_lb = i;
+                row_offset = self.slice_start - frame_start_row
+            }
+            if frame_start_row < self.slice_stop {
+                frame_idx_ub = i + 1
+            }
+            frame_start_row = buf_stop;
+        }
+
+        // then we actually decompress those frames
+        let frame_metas = &self.metadata.frames[frame_idx_lb..frame_idx_ub];
+        let total_uncompressed_size: usize = frame_metas
+            .iter()
+            .map(|meta| meta.uncompressed_size)
+            .sum::<u64>()
+            .try_into()?;
+
+        let mut decompressor = if let Some(dictionary) = &self.dictionary {
+            zstd::bulk::Decompressor::with_dictionary(dictionary)
+        } else {
+            zstd::bulk::Decompressor::new()
+        }?;
+
+        // we could make this empty initialized for better performance
+        let mut frames_values_bytes = vec![0; total_uncompressed_size];
```

Comment:
We have our own buffer to be able to easily do those unitiliazed writes. 
```suggestion
        let mut frames_values_bytes = ByteBufferMut::with_capacity_aligned(
            total_uncompressed_size,
            Alignment::new(byte_width),
        );
```
then before decompress_to_buffer we need
```
unsafe { frames_values_bytes.set_len(stop_byte) };
```
and once again at the end of the loop
```
unsafe { frames_values_bytes.set_len(total_uncompressed_size) };
```

Reply:
:+1: